// srpcd/build/builders/nodejs/nodejs.server.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodejsServerGenerator = void 0;
// do something with output
const lib_1 = require("../../lib");
const helpers_1 = require("../helpers");
function nodejsServerGenerator({ output, semantic }) {
    return __awaiter(this, void 0, void 0, function* () {
        const { map, proto_json } = yield semantic.generateProto();
        const { server } = yield semantic.generateNodeServer();
        const project_dir = new lib_1.ProjectDirectory({ dir: output });
        yield project_dir.writeFiles({ name: 'rpc.server.ts', content: server, language: 'typescript' });
        const js = yield helpers_1.compileTS(`${output}/rpc.server.ts`, {
            allowJs: true,
            declaration: true,
        });
        project_dir.writeFiles([
            { name: 'def.map.json', content: map, language: 'json' },
            { name: 'def.proto.json', content: yield proto_json(), language: 'json' },
            { name: 'rpc.server.js', content: js[1], language: 'js' },
            { name: 'rpc.server.bundle.d.ts', content: js[0], language: 'typescript' },
        ]);
        yield helpers_1.webpackBuild({
            entry: `${output}/rpc.server.js`,
            out_filename: `rpc.server.bundle.js`,
            target: 'node',
            libraryTarget: 'umd',
            bundle_output_path: output
        });
        project_dir.removeFiles(['def.map.json', 'def.proto.json', 'rpc.server.js', 'rpc.server.ts']);
    });
}
exports.nodejsServerGenerator = nodejsServerGenerator;
