// srpcd/build/examples/server/impls/DriveServiceImpl.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DriveServiceImpl = void 0;
const DriveFileImpl_1 = require("./DriveFileImpl");
const rpc_server_bundle_1 = require("../srpcd/rpc.server.bundle");
var DriveService = rpc_server_bundle_1.drive.DriveService;
class DriveServiceImpl extends DriveService {
    getDetails() {
        return Promise.resolve(new Array(2000).fill(9).map(() => ({
            id: '',
            name: 'root',
            path: 'sgeyzdfh ggvdu',
            // path: '',
            directory: true,
            size: 0,
        })));
    }
    rootFolder() {
        return __awaiter(this, void 0, void 0, function* () {
            return new DriveFileImpl_1.DriveFileImpl({
                id: '',
                name: 'root',
                path: 'yftzg',
                // path: '',
                directory: true,
                size: 0,
            });
        });
    }
}
exports.DriveServiceImpl = DriveServiceImpl;
