// srpcd/build/attributes/helpers/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReturnTypes = exports.BuiltinTypes = exports.PropertyTypes = exports.DataTypes = exports.SymbolType = exports.ClassModifier = void 0;
var ClassModifier;
(function (ClassModifier) {
    ClassModifier["SERVICE"] = "service";
})(ClassModifier = exports.ClassModifier || (exports.ClassModifier = {}));
var SymbolType;
(function (SymbolType) {
    SymbolType["FUNCTION_CALL"] = "function_call";
    SymbolType["FUNCTION"] = "function";
    SymbolType["GLOBAL"] = "block";
    SymbolType["BLOCK"] = "block";
    SymbolType["PACKAGE"] = "package";
    SymbolType["FILE"] = "file";
    SymbolType["INTERCEPTOR"] = "interceptor";
    SymbolType["VARIABLE"] = "variable";
    SymbolType["REFERENCE"] = "reference";
    SymbolType["IMPORT"] = "import";
    SymbolType["DATA"] = "data";
    SymbolType["CLASS"] = "class";
    SymbolType["STRING"] = "string";
    SymbolType["INT"] = "int";
    SymbolType["BOOLEAN"] = "boolean";
})(SymbolType = exports.SymbolType || (exports.SymbolType = {}));
exports.DataTypes = [];
exports.PropertyTypes = [
    SymbolType.DATA,
    SymbolType.STRING,
    SymbolType.INT,
    SymbolType.BOOLEAN,
];
exports.BuiltinTypes = [
    SymbolType.DATA,
    SymbolType.STRING,
    SymbolType.INT,
    SymbolType.BOOLEAN,
];
exports.ReturnTypes = [
    ...exports.PropertyTypes,
    SymbolType.CLASS,
];
